src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,238 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysis.ReadyToRun;
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+using ObjectData = ILCompiler.DependencyAnalysis.ObjectNode.ObjectData;
+
+namespace ILCompiler.ObjectWriter
+{
+    /// <summary>
+    /// Wasm object file format writer.
+    /// </summary>
+    internal sealed class WasmObjectWriter : ObjectWriter
+    {
+        public WasmObjectWriter(NodeFactory factory, ObjectWritingOptions options, OutputInfoBuilder outputInfoBuilder)
+            : base(factory, options, outputInfoBuilder)
+        {
+        }
+
+        private void EmitWasmHeader(Stream outputFileStream)
+        {
+            SectionData data = new();
+            SectionWriter headerWriter = new(this, 0, data);
+            headerWriter.Write("\0asm"u8);
+            headerWriter.Write([0x1, 0x0, 0x0, 0x0]);
+
+            data.GetReadStream().CopyTo(outputFileStream);
+        }
+
+        // TODO: for now, we are fully overriding EmitObject. As we support more features, we should
+        // see if we can re-use the base, or refactor the base method to allow for code sharing.
+        public override void EmitObject(Stream outputFileStream, IReadOnlyCollection<DependencyNode> nodes, IObjectDumper dumper, Logger logger)
+        {
+            ArrayBuilder<WasmFuncType> methodSignatures = new();
+            ArrayBuilder<string> methodNames = new();
+            ArrayBuilder<ObjectData> methodBodies = new();
+            foreach (DependencyNode node in nodes)
+            {
+                if (node is not ObjectNode)
+                {
+                    continue;
+                }
+
+                if (node is IMethodBodyNode methodNode)
+                {
+                    ObjectNode methodObject = (ObjectNode)node;
+                    methodSignatures.Add(WasmAbiContext.GetSignature(methodNode.Method));
+                    methodNames.Add(methodNode.GetMangledName(_nodeFactory.NameMangler));
+                    methodBodies.Add(methodObject.GetData(_nodeFactory));
+                    // TODO: record relocations and attached data
+                }
+            }
+
+            IEnumerable<WasmFuncType> uniqueSignatures = methodSignatures.ToArray().Distinct();
+            Dictionary<WasmFuncType, int> signatureMap = uniqueSignatures.Select((sig, i) => (sig, i)).ToDictionary();
+
+            EmitWasmHeader(outputFileStream);
+            EmitSection(() => WriteTypeSection(uniqueSignatures, logger), outputFileStream, logger);
+            EmitSection(() => WriteFunctionSection(methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
```

Comment:
```suggestion
            EmitSection(() => WriteFunctionSection(methodSignatures, signatureMap, logger), outputFileStream, logger);
```
And corresponding change in the body/signature.